File: _stats_build/statistics/most_frequent_results.rb
require_relative "../core/grouped_statistic"
require_relative "../core/events"

class MostFrequentResults < GroupedStatistic
  def initialize
    @title = "Most frequent results"
    @table_header = { "Count" => :right, "Result" => :right }
  end

  def query
    <<-SQL
      SELECT
        event_id,
        #{Database::ATTEMPTS_SUBQUERY} AS attempts
      FROM results result
      WHERE event_id != '333mbo'
    SQL
  end

  def transform(query_results)
    Events::ALL.map do |event_id, event_name|
      counts_with_results = query_results
        .select { |result| result["event_id"] == event_id }
        .flat_map do |result|
          (result["attempts"] || "").split(",").map do |v|
            { "event_id" => result["event_id"], "value" => v.to_i }
          end
        end
        .select { |result| result["value"] > 0 }
        .group_by { |result| result["value"] }
        .map { |value, results| [value, results.length] }
        .sort_by { |value, count| -count }
        .first(10)
        .map { |value, count| [count, SolveTime.new(event_id, :single, value).clock_format] }
      [event_name, counts_with_results]
    end
  end
end

File: trainer/packages/stats-build/src/statistics/most_frequent_results.ts
// NOTE: 最常出现的成绩
// 与 Ruby _stats_build/statistics/most_frequent_results.rb 1:1 对应
import { GroupedStatistic } from '../core/grouped_statistic.js';
import { EVENTS } from '../core/events.js';
import { SolveTime } from '../core/solve_time.js';
import { ATTEMPTS_SUBQUERY } from '../core/database.js';
import type { RowDataPacket } from 'mysql2';

export class MostFrequentResults extends GroupedStatistic {
  constructor() {
    super();
    this.title = 'Most frequent results';
    this.titleZh = '最常出现的成绩';
    this.tableHeader = {
      'Count': 'right',
      'Result': 'right',
    };
  }

  query(): string {
    return `
      SELECT
        event_id,
        ${ATTEMPTS_SUBQUERY} AS attempts
      FROM results result
      WHERE event_id != '333mbo'
    `;
  }

  // NOTE: 与 Ruby transform 1:1 对应——拆分 attempts、按值频率排序
  transform(rows: RowDataPacket[]): [string, unknown[][]][] {
    return Object.entries(EVENTS).map(([eventId, eventName]) => {
      // NOTE: 收集所有正值 attempt
      const valueCounts = new Map<number, number>();
      for (const row of rows) {
        if (row['event_id'] !== eventId) continue;
        const attempts = ((row['attempts'] as string) || '').split(',').map(Number);
        for (const v of attempts) {
          if (v > 0) {
            valueCounts.set(v, (valueCounts.get(v) ?? 0) + 1);
          }
        }
      }

      const results = [...valueCounts.entries()]
        .sort((a, b) => b[1] - a[1])
        .slice(0, 10)
        .map(([value, count]) => [
          count,
          new SolveTime(eventId, 'single', value).clockFormat(),
        ]);

      return [eventName, results] as [string, unknown[][]];
    });
  }
}
